piker/brokers/kucoin.py

```diff
@@ -0,0 +1,668 @@
+        self._key_id: str
+        self._key_secret: str
+        self._key_passphrase: str
+        self._authenticated: bool = False
+
+        config: BrokerConfig | None = get_config()
+
+        if config and config.key_id and config.key_secret and config.key_passphrase:
+            self._authenticated = True
+            self._key_id = config.key_id
+            self._key_secret = config.key_secret
+            self._key_passphrase = config.key_passphrase
+            log.info('User credentials added')
+
+    def _gen_auth_req_headers(
+        self,
+        action: Literal['POST', 'GET'],
+        endpoint: str,
+        api_v: str = 'v2',
+    ) -> dict[str, str | bytes]:
+        '''
+        Generate authenticated request headers
+        https://docs.kucoin.com/#authentication
+
+        '''
+        now = int(time.time() * 1000)
+        path = f'/api/{api_v}{endpoint}'
+        str_to_sign = str(now) + action + path
+
+        signature = base64.b64encode(
+            hmac.new(
+                self._key_secret.encode('utf-8'),
+                str_to_sign.encode('utf-8'),
+                hashlib.sha256,
+            ).digest()
+        )
+
+        passphrase = base64.b64encode(
+            hmac.new(
+                self._key_secret.encode('utf-8'),
+                self._key_passphrase.encode('utf-8'),
+                hashlib.sha256,
+            ).digest()
+        )
+
+        return {
+            'KC-API-SIGN': signature,
+            'KC-API-TIMESTAMP': str(now),
+            'KC-API-KEY': self._key_id,
+            'KC-API-PASSPHRASE': passphrase,
+            # XXX: Even if using the v1 api - this stays the same
+            'KC-API-KEY-VERSION': '2',
+        }
+
+    async def _request(
+        self,
+        action: Literal['POST', 'GET'],
+        endpoint: str,
+        api_v: str = 'v2',
+        headers: dict = {},
+    ) -> Any:
+        '''
+        Generic request wrapper for Kucoin API
+
+        '''
+        if self._authenticated:
+            headers = self._gen_auth_req_headers(action, endpoint, api_v)
+
+        api_url = f'https://api.kucoin.com/api/{api_v}{endpoint}'
+
+        res = await asks.request(action, api_url, headers=headers)
+
+        if 'data' in res.json():
+            return res.json()['data']
+        else:
+            log.error(f'Error making request to {api_url} -> {res.json()["msg"]}')
+            return res.json()['msg']
+
+    async def _get_ws_token(self, private: bool = False) -> tuple[str, int] | None:
+        '''
+        Fetch ws token needed for sub access
+
+        '''
+        token_type = 'private' if private else 'public'
+        data: dict[str, Any] | None = await self._request(
+            'POST', f'/bullet-{token_type}', 'v1'
+        )
+
+        if data and 'token' in data:
+            ping_interval: int = data['instanceServers'][0]['pingInterval']
+            return data['token'], ping_interval
+        elif data:
```

Comment:
what's normally happening in this case?

Reply:
so if `bool(data)` being false-y should never happen yah?

cuz in that case this will exit silently i think?

Reply:
Let's account for that just in case. If the API request is malformed this could happen.